shared/helper/adm_preset_definitions_helper.cpp

```diff
@@ -0,0 +1,608 @@
+#include "adm_preset_definitions_helper.h"
+#include <adm/common_definitions.hpp>
+#include <adm/parse.hpp>
+#include <algorithm>
+#include <string>
+#include <string_view>
+#include <sstream>
+#include <vector>
+#include <helper/container_helpers.hpp>
+#include <helper/supplementary_definitions.hpp>
+#include <helper/cartesianspeakerlayouts.h>
+
+namespace {
+	std::vector<std::string_view> splitString(const std::string& input, char delimiter, size_t maxElements) {
+		std::vector<std::string_view> result;
+		std::string_view str(input);
+		size_t start = 0;
+
+		for (size_t i = 0; i < maxElements - 1; ++i) {
+			size_t end = str.find(delimiter, start);
+			if (end == std::string_view::npos) {
+				break;
+			}
+			result.push_back(str.substr(start, end - start));
+			start = end + 1;
+		}
+
+		// Add the remaining text, even if it contains more delimiters
+		result.push_back(str.substr(start));
+
+		return result;
+	}
+
+	template <typename Prop, typename Elm>
+	bool matchingProp(const Elm& elmA, const Elm& elmB) {
+		bool hasProp = elmA.template has<Prop>();
+		if (hasProp != elmB.template has<Prop>())
+			return false;
+		if (hasProp) {
+			auto valA = elmA.template get<Prop>();
+			auto valB = elmB.template get<Prop>();
+			if (valA != valB) return false;
+		}
+		return true;
+	}
+}
+
+AdmPresetDefinitionsHelper::AdmPresetDefinitionsHelper()
+{
+}
+
+AdmPresetDefinitionsHelper::~AdmPresetDefinitionsHelper()
+{
+}
+
+std::shared_ptr<AdmPresetDefinitionsHelper> AdmPresetDefinitionsHelper::getSingleton()
+{
+	static auto instance = std::make_shared<AdmPresetDefinitionsHelper>();
+	return instance;
+}
```

Comment:
Yeah, it's confusing - partly as `static` is used to mean two different things in this context, and partly as we're only doing it to avoid the [static order initialisation fiasco](https://en.cppreference.com/w/cpp/language/siof) which is a confusing byproduct of how C/C++ compilation works, so the intent isn't particularly clear.

The `static` at the start of the member function declaration, just means the function is associated with the class rather than an instance of the class. So it's namespaced using the class name, and it can have access to private/protected things on the class, but has no access to anything that depends on a particular instance of the class (normal member variables/ member functions). 
An example you sometimes see where the access rights are useful is something like
```c++
class S {
  private:
   S()
  public:
  static std::shared_ptr<S> create() { return std::make_shared<S>() );
}
```
for when you want to guarantee something is only ever used via a shared_ptr (it's in libadm's Document class for example)

The `static` inside the function, for the variable declaration, means that the variable is initialised the first time the fucntion is called, and lives for the duration of the program. The advantage over global or class member static variables is that the initialisation is guaranteed to happen when the function is first called rather than being implementation defined, so you don't run into any weird problems. I think the definition of 'first' is also thread safe as of some recent-ish version of the standard (my googling failed me today). As the lifetime of a function static variable is the lifetime of the whole program, you can return a dangling reference safely, which would otherwise be a definite bad thing.

I think the idiomatic way would be
```c++
AdmPresetDefinitionsHelper const& AdmPresetDefinitionsHelper::getSingleton()
{
	static auto* instance = new<AdmPresetDefinitionsHelper>();
	return *instance;
}
```

By using new, you don't have to worry about destructor ordering (which isn't defined) in case of any inter-dependencies - the destructor never gets called. This has tripped me up before. Like returing a dangling reference, leaving a new hanging there feels bad (memory leaks!), but also doesn't matter - I forget who to attribute it to, but there's a quote along the lines of "if the house is burning down you don't need to stop to sweep the floor", which references worrying about destroying things neatly when a process exits. The OS is going to reclaim all the resources unless you're doing something weird (e.g. blocking on something) - you can't really 'leak' a resource that's got the same lifetime as the process.

```c++
AdmPresetDefinitionsHelper const& AdmPresetDefinitionsHelper::getSingleton()
{
	static auto instance = new<AdmPresetDefinitionsHelper>();
	return instance;
}
```
Would be fine if you *were* doing something weird, but you'd have to exercise extra care about any inter-dependencies between static objects. Similarly, the shared_pointer might actually hurt here, as it will call the destructor.